linera-faucet/server/src/lib.rs

```diff
@@ -118,17 +123,57 @@ where
     C: ClientContext,
 {
     async fn do_claim(&self, owner: AccountOwner) -> Result<ClaimOutcome, Error> {
-        let client = self.context.lock().await.make_chain_client(self.chain_id)?;
+        let main_client = self
+            .context
+            .lock()
+            .await
+            .make_chain_client(self.main_chain_id)?;
+        let maybe_tmp_chain_id = *self.tmp_chain_id.lock().await;
+        let main_balance = main_client.local_balance().await?;
+        let tmp_chain_id = match maybe_tmp_chain_id {
+            Some(tmp_chain_id) => tmp_chain_id,
+            None => {
+                let key_pair = main_client.key_pair().await?;
+                let balance = self
+                    .amount
+                    .try_add(MAX_FEE)?
+                    .try_mul(u128::from(self.max_claims_per_chain))?
+                    .try_add(MAX_FEE)? // One more block fee for closing the chain.
+                    .min(main_balance.try_sub(MAX_FEE)?);
+                let ownership = main_client.chain_state_view().await?.ownership().clone();
+                let (message_id, certificate) = main_client
+                    .open_chain(ownership, ApplicationPermissions::default(), balance)
+                    .await?
+                    .try_unwrap()?;
+                let chain_id = ChainId::child(message_id);
+                info!("Switching to a new faucet chain {chain_id:8}");
+                self.context
+                    .lock()
+                    .await
+                    .update_wallet_for_new_chain(
+                        chain_id,
+                        Some(key_pair),
+                        certificate.block().header.timestamp,
+                    )
+                    .await?;
+                *self.tmp_chain_id.lock().await = Some(chain_id);
+                chain_id
```

Comment:
I'd extract this `None => { ... }` clause to a method - `fn open_new_faucet_chain`.